aten/src/ATen/DeviceGuard.h

```diff
@@ -6,42 +6,55 @@
 #include <ATen/detail/CUDAHooksInterface.h>
 #include <c10/util/Exception.h>
 #include "c10/util/Optional.h"
+#include <c10/detail/DeviceGuardImplInterface.h>
 
 #include <cstddef>
 
 namespace at {
 /// RAII guard that sets a certain default device in its constructor, and
-/// changes it back to the device that was originally active upon destruction.
+/// changes it back to the device (for that device type) that was originally
+/// active upon destruction.
 ///
-/// The device is always reset to the one that was active at the time of
-/// construction of the guard. Even if you `set_device` after construction, the
-/// destructor will still reset the device to the one that was active at
-/// construction time.
-struct DeviceGuard {
-  /// Default constructor, does nothing.
-  DeviceGuard() = default;
-
+/// If the device is changed via this guard to a different one than the
+/// active one at construction time, this guard will reset it to the one
+/// that was active at the time of construction of the guard.  WARNING: if
+/// you change the current device out-of-band, e.g., by directly calling
+/// cudaSetDevice(), DeviceGuard is NOT guaranteed to reset it upon
+/// exiting this scope.  The contract required by DeviceGuard is that inner code
+/// leaves the device in the same state that DeviceGuard set it.  In DEBUG mode,
+/// we check for this invariant.
+///
+/// If a DeviceGuard is constructed without specifying a device type (this
+/// can occur if you, e.g., pass a nullopt to the constructor), it behaves as if
+/// it were a no-op "CPU" guard; e.g., current_device() reports that the current
+/// device is kCPU.  This is different from passing Device(kCUDA, -1), which
```

Comment:
if we take an optional<Device> here...why don't we return an optional<Device>?

Reply:
In this formulation, `DeviceGuard g(nullopt)` is defined to be `DeviceGuard g(kCPU)`, so a device guard always has a DeviceType. I agree it could be done the other way. I ended up doing it this way mostly because avoiding storing an optional member avoids the lgenfe bug (https://github.com/pytorch/pytorch/issues/12117) which, while not insoluble, is a little less friendly.

Reply:
I don't think we should do it this way.  We had a ton of magic mapping from "optional" devices to CPU before we started using optional and it just led to bugs, confusion, etc.  The error linked can also be easily worked around and shouldn't have a transitive effect given it's inside a .cu file, where you probably don't have (non-cuda) derived DeviceGuards.